In [86]:
import pandas as pd
import numpy as np
import datetime
import warnings
import datetime
from sklearn.linear_model import LinearRegression
from sklearn.decomposition import PCA
from scipy.optimize import minimize

import matplotlib.pyplot as plt
%matplotlib inline
plt.rcParams['figure.figsize'] = (12,6)
plt.rcParams['font.size'] = 15
plt.rcParams['legend.fontsize'] = 13

from matplotlib.ticker import (MultipleLocator,
                               FormatStrFormatter,
                               AutoMinorLocator)

In [87]:
#1.1
treasury_bond = []
swap_bond = []
pv_treasury_bond = []
pv_swap_bond = []
pvt_treasury_bond = []
pvt_swap_bond = []
discount = 1
for i in range(0, 60):
    discount = discount * (1+ 0.04193/2)
    if i != 59:
        treasury_bond.append(2.125)
        swap_bond.append(2.128)
        pv_treasury_bond.append(2.125/discount)
        pv_swap_bond.append(2.128/discount)
        pvt_treasury_bond.append(2.125/discount*(i+1)/2)
        pvt_swap_bond.append(2.128/discount*(i+1)/2)
    else:
        treasury_bond.append(102.125)
        swap_bond.append(102.128)
        pv_treasury_bond.append(102.125/discount)
        pv_swap_bond.append(102.128/discount)
        pvt_treasury_bond.append(102.125/discount*(i+1)/2)
        pvt_swap_bond.append(102.128/discount*(i+1)/2)

In [88]:
treasury_bond[1]-swap_bond[1]

-0.0030000000000001137

In [89]:
#1.2
durationtreasury = sum(pvt_treasury_bond)/sum(pv_treasury_bond)
durationswap = sum(pvt_treasury_bond)/sum(pvt_swap_bond)-30
durationtreasury

17.288293145479983

In [90]:
dollar_durationtreasury = durationtreasury*sum(pv_treasury_bond)
dollar_durationtreasury

1745.563315728143

In [91]:
durationswap

-29.000712547007033

In [92]:
dollar_durationswap = durationswap*sum(pv_treasury_bond)
dollar_durationswap

-2928.142155275034

so the duration is negative and its manititude is greater than duration of tresury bond.

In [97]:
#1.3
pos = -500 * durationswap/durationtreasury
pos

838.7384544838442

so we are expected to long 838.74 million of treasury bonds.

1.4
If the spread is negative, then we are expected to have treasury part increases, but swap part decreases more than the gain of swap.

In [94]:
#1.5
pv_treasury_bond_new = []
pv_swap_bond_new = []
pv_treasury_bond_new.append(pv_treasury_bond[1])
pv_swap_bond_new.append(pv_swap_bond[1])
discount = 1
for i in range(1, 60):
    discount = discount * (1+ 0.0436/2)
    if i != 59:
        pv_treasury_bond_new.append(2.125/discount)
        pv_swap_bond_new.append(2.04/discount)
    else:
        pv_treasury_bond_new.append(102.125/discount)
        pv_swap_bond_new.append(102.04/discount)

In [95]:
sum(pv_treasury_bond_new)

100.22252389820919

In [96]:
sum(pv_swap_bond_new)

97.41870112169434

In [102]:
#1.6
loss = 500*sum(pv_treasury_bond_new)/100+pos*sum(pv_swap_bond_new)/100-500-pos
loss

-20.537726826463768

In [103]:
-(500+pos)*0.02+loss

-47.31249591614065